layerindex/utils.py

```diff
@@ -253,33 +290,53 @@ def parse_layer_conf(layerdir, data, logger=None):
     data = parse_conf(conf_file, data)
     data.expandVarref('LAYERDIR')
 
-def runcmd(cmd, destdir=None, printerr=True, logger=None):
+child_pid = 0
+def runcmd(cmd, destdir=None, printerr=True, outfile=None, logger=None, shell=False):
     """
         execute command, raise CalledProcessError if fail
         return output if succeed
     """
     if logger:
         logger.debug("run cmd '%s' in %s" % (cmd, os.getcwd() if destdir is None else destdir))
-    out = tempfile.TemporaryFile()
+    if outfile:
+        out = open(outfile, 'wb+')
+    else:
+        out = tempfile.TemporaryFile()
+
+    def onsigusr2(sig, frame):
+        # Kill the child process
+        os.kill(child_pid, signal.SIGTERM)
+    signal.signal(signal.SIGUSR2, onsigusr2)
     try:
-        subprocess.check_call(cmd, stdout=out, stderr=out, cwd=destdir, shell=True)
-    except subprocess.CalledProcessError as e:
+        proc = subprocess.Popen(cmd, stdout=out, stderr=out, cwd=destdir, shell=shell)
+        global child_pid
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Global variables are dangerous because they can be simultaneously accessed from multiple sections of a program. This frequently results in bugs. Most bugs involving global variables arise from one function reading and acting on the value of a global variable before another function has the chance to set it to an appropriate value. We recommend using a configuration module to mutate global state.

[Learn more](https://docs.python.org/3/faq/programming.html#how-do-i-share-global-variables-across-modules)